test/e2e/e2e_test.go

```diff
@@ -851,7 +863,17 @@ func testE2E() {
 			"reason=VolumeResizeFailed"
 		var events corev1.EventList
 		err = getObjects(&events, "events", "-n", ns, "--field-selector="+fieldSelector)
-		Expect(err).To(BeEquivalentTo(ErrObjectNotFound))
+		wasModified := "the object has been modified; please apply your changes to the latest version and try again"
```

Comment:
Even though this test is problematic right now I believe it is useful. That is because every failed run of the csi-resizer will be reported in such an event. Especially this is useful because retries are spotted here that should not happen in the first place. E.g. here, no one would expect any errors because there is nothing that should trigger a retry.

If we change our stance to say all failures are okay, then thats fine for me.

Reply:
This test ensures no event occurs before the test expects events.
https://github.com/topolvm/topolvm/pull/903/files#diff-d5d12b6ea07b0a78c9ad1987e1486c565ecf16839f77a93e91b09475cf6c0e99L856-R895

BTW, can we check if the resize fails by checking the PVC status?
https://github.com/topolvm/topolvm/pull/903/files#diff-d5d12b6ea07b0a78c9ad1987e1486c565ecf16839f77a93e91b09475cf6c0e99R878-R881

Reply:
Sure it would be possible to check for PersistentVolumeClaimControllerResizeFailed or PersistentVolumeClaimNodeResizeFailed conditions to indicate PVC resize failure

Reply:
I'd like to write it the way you suggested. What do you think? @daichimukai

Reply:
added the test case as suggested

Reply:
Thanks to the new checking code, we can now remove the event check codes, which are redundant and difficult to understand the intention to check resize failure. Could you remove them?

Reply:
The event check is meant to check for explicit failures after going over capacity. The event checking code is tested before that, when a succesful resize is expected. We do not expect any failure events during this, which is why we have this test. I suggest we keep the test function in for that reason.

Reply:
I think what we want to check is that the PV resizing succeeds. This can be achieved by checking no events, but it is an implicit condition, IMO. Therefore, I prefer to use explicit checking and delete implicit one.

This is a side note, we ignore events showing retry on conflicts [here](https://github.com/topolvm/topolvm/pull/903/files#diff-d5d12b6ea07b0a78c9ad1987e1486c565ecf16839f77a93e91b09475cf6c0e99R877). If there are such events, [this condition](https://github.com/topolvm/topolvm/pull/903/files#diff-d5d12b6ea07b0a78c9ad1987e1486c565ecf16839f77a93e91b09475cf6c0e99R896-R907) is accidentally met.

Reply:
I think we have 2 scenarios: successful resize, and failinig resize due to lack of capacity.

For successful resize, having no failure events might be seen as critical or not. Of course in an eventually consistent system we are fine with failures every now and then but if we have a system in which we can ensure optimal conditions, then testing that no intermediate failure event occurs makes sense imo.

For failing resize, similarly it makes sense to verify that we have events checked that report this failure on top of the failing PVC status.

That being said, Im also fine with removing after all if you think thats not enough of a reason, not married to this idea.

Reply:
I found the checks were introduced in [this commit](https://github.com/topolvm/topolvm/pull/101/commits/9f84e16f214d94db0f67fae58f098a74bff7e347) and asked the authors about the intention. The answer is that the first check is a precondition of the failing resize to ensure no events occur and the last is a postcondition to confirm events happen. They also told me that checking events is just a tool to check the failing resize, alternatives could be used.

I feel checking the status is more readable, so I'd like to use it. However, I found this method can not be used because  external-resizer sets the status only when the RecoverVolumeExpansionFailure feature gate is enabled (default false now).
https://github.com/kubernetes-csi/external-resizer/blob/master/pkg/controller/controller.go#L336
You can see [expandAndRecover](https://github.com/kubernetes-csi/external-resizer/blob/master/pkg/controller/expand_and_recover.go#L28) is called, and it calls [callResizeOnPlugin](https://github.com/kubernetes-csi/external-resizer/blob/master/pkg/controller/expand_and_recover.go#L156), and then it calls [markControllerExpansionFailed](https://github.com/kubernetes-csi/external-resizer/blob/master/pkg/controller/resize_status.go#L90) which updates the status.

But, why did our test succeed? It is because we missed `.Should(Succeed())` [here](https://github.com/topolvm/topolvm/pull/903/files#diff-d5d12b6ea07b0a78c9ad1987e1486c565ecf16839f77a93e91b09475cf6c0e99R928).

So, we should retain the event check code and remove the status check.

Reply:
Thanks for going in detail on this and explaining that for me. Switched back to events for now with the exception for the Collision to avoid flakes.